plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/functions/tablechanges/TableChangesFunctionProcessor.java

```diff
@@ -0,0 +1,173 @@
+        org.apache.iceberg.types.Type[] partitionColumnTypes = partitionSpec.fields().stream()
+                .map(field -> field.transform().getResultType(tableSchema.findType(field.sourceId())))
+                .toArray(org.apache.iceberg.types.Type[]::new);
+
+        int delegateColumnIndex = 0;
+        int[] delegateColumnMap = new int[functionHandle.getColumns().size()];
+        Optional<Integer> changeTypeIndex = Optional.empty();
+        Optional<Integer> changeVersionIndex = Optional.empty();
+        Optional<Integer> changeTimestampIndex = Optional.empty();
+        Optional<Integer> changeOrdinalIndex = Optional.empty();
+        for (int columnIndex = 0; columnIndex < functionHandle.getColumns().size(); columnIndex++) {
+            IcebergColumnHandle column = functionHandle.getColumns().get(columnIndex);
+            if (column.getId() == DATA_CHANGE_TYPE_ID) {
+                changeTypeIndex = Optional.of(columnIndex);
+                delegateColumnMap[columnIndex] = -1;
+            }
+            else if (column.getId() == DATA_CHANGE_VERSION_ID) {
+                changeVersionIndex = Optional.of(columnIndex);
+                delegateColumnMap[columnIndex] = -1;
+            }
+            else if (column.getId() == DATA_CHANGE_TIMESTAMP_ID) {
+                changeTimestampIndex = Optional.of(columnIndex);
+                delegateColumnMap[columnIndex] = -1;
+            }
+            else if (column.getId() == DATA_CHANGE_ORDINAL_ID) {
+                changeOrdinalIndex = Optional.of(columnIndex);
+                delegateColumnMap[columnIndex] = -1;
+            }
+            else {
+                delegateColumnMap[columnIndex] = delegateColumnIndex;
+                delegateColumnIndex++;
+            }
+        }
+
+        this.split = split;
+        this.pageSource = icebergPageSourceProvider.createPageSource(
+                session,
+                functionHandle.getColumns(),
+                tableSchema,
+                partitionSpec,
+                PartitionData.fromJson(split.getPartitionDataJson(), partitionColumnTypes),
+                ImmutableList.of(),
+                DynamicFilter.EMPTY,
+                TupleDomain.all(),
+                split.getPath(),
+                0,
+                split.getLength(),
+                split.getFileSize(),
+                split.getFileRecordCount(),
+                split.getPartitionDataJson(),
+                split.getFileFormat(),
+                functionHandle.getNameMappingJson().map(NameMappingParser::fromJson));
+        this.delegateColumnMap = delegateColumnMap;
+        this.changeTypeIndex = changeTypeIndex;
+        this.changeVersionIndex = changeVersionIndex;
+        this.changeTimestampIndex = changeTimestampIndex;
+        this.changeOrdinalIndex = changeOrdinalIndex;
+    }
+
+    @Override
+    public TableFunctionProcessorState process()
+    {
+        while (!pageSource.isFinished()) {
```

Comment:
I assume it would make sense if you follow this @findepi 's advise https://github.com/trinodb/trino/pull/17999#discussion_r1238215794